### PR TITLE
Yet another login method(without code)

### PR DESCRIPTION
**Pull request by unasuke**
I'm ruboty and jobcan user.
But, I don't know my jobcan code.

So, I implemented yet another login method.
That use client id, email address and password.

This PR keep compatibility maybe...

### CODE REVIEW

File: lib/ruboty/actions/jobcan.rb [hunk 1]
```diff
@@ -55,6 +55,14 @@ def user_data
         brain_space[message.from_name] ||= {}
       end
 
+      def present_login_env?
+        !(
+          ENV["RUBOTY_JOBCAN_CLIENT_ID"].nil? &&
+          ENV["RUBOTY_JOBCAN_EMAIL"].nil?     &&
+          ENV["RUBOTY_JOBCAN_PASSWORD"].nil?
+        )
+      end
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb:100]: `present_login_env?` returns true even if any of the env vars are unset. It should be

``` ruby
def present_login_env?
  !!(
    ENV["RUBOTY_JOBCAN_CLIENT_ID"] &&
    ENV["RUBOTY_JOBCAN_EMAIL"] &&
    ENV["RUBOTY_JOBCAN_PASSWORD"]
  )
end
```

and the name should be `login_env_present?`

File: lib/ruboty/actions/jobcan.rb [hunk 2]
```diff
@@ -24,9 +24,11 @@ def remember_group_id
       end
 
       def punch_clock(in_out = :auto)
-        unless (code = user_data["code"])
+        if !present_login_env? && !(code = user_data["code"])
           message.reply("I don't know your JOBCAN code.")
           return
+        else
+          code = nil
         end
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb]: `code` always be set to nil along the happy path. This if-else should be rewritten as:

``` ruby
unless (code = user_data["code"]) || present_login_env?
  message.reply("I don't know your JOBCAN code.")
  return
end
```

File: lib/ruboty/handlers/jobcan.rb [hunk 3]
```diff
@@ -26,17 +26,35 @@ class Jobcan < Base
       )
 
       on(
+        /punch the clock at (?<group_name>.+)\z/,
+        description: "Punch the clock on JOBCAN with gourp alias",
```
Comment (uasi) [inline lib/ruboty/handlers/jobcan.rb]: s/gourp/group/

File: lib/ruboty/handlers/jobcan.rb [hunk 4]
```diff
@@ -26,17 +26,35 @@ class Jobcan < Base
...
+
+      on(
         /clock in\z/,
         description: "Clock in on JOBCAN",
         name: "clock_in",
       )
 
       on(
+        /clock in at (?<group_name>.+)\z/,
+        description: "Clock in on JOBCAN with gourp alias",
```
Comment (uasi) [inline lib/ruboty/handlers/jobcan.rb]: s/gourp/group/

File: lib/ruboty/handlers/jobcan.rb [hunk 5]
```diff
@@ -26,17 +26,35 @@ class Jobcan < Base
...
+
+      on(
         /clock out\z/,
         description: "Clock out on JOBCAN",
         name: "clock_out",
       )
 
+      on(
+        /clock out at (?<group_name>.+)\z/,
+        description: "Clock out on JOBCAN with gourp alias",
```
Comment (uasi) [inline lib/ruboty/handlers/jobcan.rb]: s/gourp/group/

File: lib/ruboty/actions/jobcan.rb [hunk 6]
```diff
@@ -70,6 +74,10 @@ def present_login_env?
         )
       end
 
+      def exist_group_alias?(group_name)
+        !!user_data["alias_#{group_name}"]
+      end
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb:104]: should be `group_alias_exist?`

File: README.md [hunk 7]
```diff
@@ -1,27 +1,56 @@
...
-e.g. `http://jobcan.jp/m?code=YOUR_CODE_GOES_HERE`.
-Check out the invitation mail from JOBCAN for the URL.
-
 `<group_id>` is the numeric ID of your group;
 ask your manager or check out the source code of https://ssl.jobcan.jp/employee for it.
 
 ![group ID](img/group_id.png)
 
+If you want to use different group id, you can use that.
+```
```
Comment (uasi) [inline README.md:44]: Insert a blank line between "If you want..." and the code block.

File: lib/ruboty/actions/jobcan.rb [hunk 8]
```diff
@@ -28,17 +28,22 @@ def register_group_alias
...
-          message.reply("I don't know your JOBCAN group ID.")
-          return
-        end
+
+        group_id = if at
+                     user_data["alias_#{at}"]
+                   elsif !user_data["group_id"].nil?
+                     user_data["group_id"]
+                   else
+                     message.reply("I don't know your JOBCAN group ID.")
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb]: `return` here.

File: lib/ruboty/actions/jobcan.rb [hunk 9]
```diff
@@ -23,48 +23,105 @@ def remember_group_id
...
         message.reply("Error: #{$!}")
       end
 
+      def punch_clock_at
+        unless exist_group_alias?(message[:group_name])
+          cannot_find_alias_message(message[:group_name])
+          return
+        end
+        punch_clock(:auto, at: message[:group_name])
+      end
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb:57]: I would simply pass the given group name to `punch_clock`

``` ruby
def punch_clock_at
  punch_clock(:auto, at: message[:group_name])
end
```

and let `puch_clock` resolve alias

``` ruby
def punch_clock(in_out = :auto, at = nil)
  ...
  group_id = if at
               user_data["alias_#{at}"] ||
                 return message.reply("I don't know your JOBCAN group alias '#{at}'.")
             else
               user_data["group_id"] ||
                 return message.reply("I don't know your JOBCAN group ID.")
             end
  ...
end
```

**Comment by unasuke** [comment]
@uasi I've fixed. Thanks.

File: lib/ruboty/actions/jobcan.rb [hunk 10]
```diff
@@ -23,48 +23,101 @@ def remember_group_id
...
+
+        group_id = if at
+                     user_data["alias_#{at}"] or
+                       return cannot_find_alias_message(at)
+                   elsif !user_data["group_id"].nil?
+                     user_data["group_id"]
+                   else
+                     user_data["group_id"] or
+                       return message.reply("I don't know your JOBCAN group ID.")
+                   end
```
Comment (uasi) [inline lib/ruboty/actions/jobcan.rb:45]: The elsif block and the else block do the same thing -- both tries to fetch `user_data["group_id"]`. The elsif block could be removed.

Comment (uasi) [inline lib/ruboty/actions/jobcan.rb:45]: I'll fix that later.

**Comment by uasi** [comment]
@unasuke Great! Looks almost good to me.

**Comment by uasi** [comment]
@unasuke Thanks!

**Comment by unasuke** [comment]
@uasi Thank you too!!

**Pull request merged**